### PR TITLE
`web3-eth` `sendTransaction` and `sendSignedTransaction` `PromiEvent` tests #4810

### PR DESCRIPTION
**Pull request by Spacesai1or**
### What This PR Does

- Adds `data` property to `BaseTransaction` type in `packages/web3-common/src/eth_execution_api.ts`
    - There is a [PR](https://github.com/ethereum/execution-apis/pull/201) discussing why
- Updated `sendTransaction` and `sendSignedTransaction` to use `ReceiptInfoFormatted` instead of `ReceiptInfo`
    - Updated `packages/web3-eth-ens/src/ens.ts` accordingly
- Added [SendTransactionOptions](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/src/types.ts#L266) to `sendTransaction`
- Init `packages/web3-eth/test/unit/rpc_method_wrappers/fixtures/sendTransaction.ts`
- Init `packages/web3-eth/test/unit/rpc_method_wrappers/sendTransaction.test.ts`
- Init `packages/web3-eth/test/unit/rpc_method_wrappers/fixtures/sendSignedTransaction.ts`
- Init `packages/web3-eth/test/unit/rpc_method_wrappers/sendSignedTransaction.test.ts`
- Init `packages/web3-eth/src/utils/wait_for_transaction_receipt.ts`
- Init `packages/web3-eth/src/utils/watch_transaction_for_confirmations.ts`

### Points of Discussion

- [x] Regarding [sendTransaction.test.ts](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/sendTransaction.test.ts) and [sendTransaction.ts fixture](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/fixtures/sendTransaction.ts), I'm proposing a refactor of at least how the `web3-eth` tests are done to make it more manageable

    Key changes:
    - Instead of having files such as `web3_rpc_method_wrappers_no_parameters.test.ts`, `web3_rpc_method_wrappers_with_parameters.test.ts`, etc. I'd like to split the file related tests into directories:
        - `web3_rpc_method_wrappers_no_parameters.test.ts` and `web3_rpc_method_wrappers_with_parameters.test.ts` become `web3-eth/test/unit/rpc_method_wrappers/`
        - `rpc_methods_no_parameters.test.ts` and `rpc_methods_with_parameters.test.ts` become `web3-eth/test/unit/rpc_methods/`
        - `web3_eth_methods_no_parameters.test.ts` and `web3_eth_methods_with_parameters.test.ts` become `web3-eth/test/unit/web3_eth/`
    - Instead of having huge files that contains all the tests for a file (e.g. [web3_rpc_method_wrappers_with_parameters.test.ts](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-eth/test/unit/web3_rpc_method_wrappers_with_parameters.test.ts)), we split each method test into it's own file (e.g. [rpc_method_wrappers/sendTransaction.ts](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/sendTransaction.test.ts))
    - Instead of having huge fixture files that contain all the test data for every test (e.g. [rpc_methods_wrappers](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-eth/test/fixtures/rpc_methods_wrappers.ts)), we have individual test fixture files (e.g. [rpc_method_wrappers/fixtures/sendTransaction.ts](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/fixtures/sendTransaction.ts))
        - We'll probably end up having a `shared.ts` file that contains data used across tests e.g. [inputTransaction](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/fixtures/sendTransaction.ts#L4) and [expectedReceiptInfo](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/fixtures/sendTransaction.ts#L19)

    The purpose of these changes are to:

    - Make the test/fixture files easier to navigate and single purposed (each test file is testing only one specific method, each fixture file only contains data for a specific method (similar to how it was done in [1.x](https://github.com/ChainSafe/web3.js/tree/1.x/test)))
    - The way majority of our tests are written where each function gets a singular [it.each](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/web3_rpc_method_wrappers_with_parameters.test.ts#L56) makes is very difficult to determine what went wrong when a test fails and multiple `expects` are included in a test (you end up having to comment out `expects` until you find the failing one). In the [purposed](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/test/unit/rpc_method_wrappers/sendTransaction.test.ts) architecture, some code is repeated, but it becomes explicitly clear what `expect` failed. It also allows for easier testing of multiple scenarios when testing a method affected by different `Web3Context` configs or method options (e.g. [ignoreGasPricing](https://github.com/ChainSafe/web3.js/blob/55409e96590b44b93490c4152d86fa760a3c634e/packages/web3-eth/src/rpc_method_wrappers.ts#L394) in `sendTransaction`)

### CODE REVIEW

**Comment by nazarhussain** [comment]
Seems good to me, but some of the change (files) may conflict with the PR I am working on. We will sort it it out later.

**Review by avkos — APPROVED** [review]

File: packages/web3-eth/src/rpc_method_wrappers.ts [hunk 1]
```diff
@@ -453,42 +362,50 @@ export function sendTransaction(
 	return promiEvent;
 }
 
-export const sendSignedTransaction = async (
+export const sendSignedTransaction = (
```
Comment (jdevcs) [inline packages/web3-eth/src/rpc_method_wrappers.ts]: [this](https://github.com/ChainSafe/web3.js/blob/a09482b0eeab653f56a2147fd236a643688e68a2/packages/web3-eth/src/web3_eth.ts#L163) also needs to be updated, it should be using current sendSignedTransaction

**Review by jdevcs — CHANGES_REQUESTED** [review]

File: packages/web3-eth/src/types.ts [hunk 2]
```diff
@@ -262,22 +255,27 @@ export type SendTransactionEvents = {
...
 		latestBlockHash: HexString32Bytes;
 	};
 };
 
+export interface SendTransactionOptions {
+	returnType?: ValidTypes;
+	ignoreGasPricing?: boolean;
+}
+
 export type SendSignedTransactionEvents = {
```
Comment (jdevcs) [inline packages/web3-eth/src/types.ts]: There is same promise combined event emitter for `sendTransaction` and `sendSignedTransaction`
```suggestion
export type SendTransactionEvents = {
```
could you make these rename changes in other parts of code as well.

**Review by jdevcs — CHANGES_REQUESTED** [review]

File: packages/web3-common/src/eth_execution_api.ts [hunk 3]
```diff
@@ -43,8 +43,9 @@ export interface BaseTransaction {
 	readonly nonce: Uint;
 	readonly gas: Uint;
 	readonly value: Uint;
-	// TODO - Investigate if this should actually be data instead of input
+	// TODO - https://github.com/ethereum/execution-apis/pull/201
```
Comment (jdevcs) [inline packages/web3-common/src/eth_execution_api.ts:46]: Good work

File: packages/web3-eth/src/utils/watch_transaction_for_confirmations.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	transactionReceipt: ReceiptInfo,
+) {
+	if (
+		transactionReceipt === undefined ||
+		transactionReceipt === null ||
+		transactionReceipt.blockHash === undefined ||
+		transactionReceipt.blockHash === null
+	)
+		throw new TransactionMissingReceiptOrBlockHashError({
+			receipt: transactionReceipt,
```
Comment (jdevcs) [inline packages/web3-eth/src/utils/watch_transaction_for_confirmations.ts:29]: if transactionReceipt is `undefined` or `null`, will be difficult to debug this error so good to pass [transactionHash](https://github.com/ChainSafe/web3.js/blob/a09482b0eeab653f56a2147fd236a643688e68a2/packages/web3-eth/src/rpc_method_wrappers.ts#L377) to `watchTransactionForConfirmations` [here](https://github.com/ChainSafe/web3.js/blob/a09482b0eeab653f56a2147fd236a643688e68a2/packages/web3-eth/src/rpc_method_wrappers.ts#L401) and show that in detail of error.

**Review by jdevcs — CHANGES_REQUESTED** [review]

File: packages/web3-eth/src/rpc_method_wrappers.ts (hunk 1 above)
Comment (Spacesai1or) [reply packages/web3-eth/src/rpc_method_wrappers.ts]: `sendRawTransaction` is the name of the RPC method, Web3.js just aliases it as `sendSignedTransaction`

File: packages/web3-eth/src/types.ts (hunk 2 above)
Comment (Spacesai1or) [reply packages/web3-eth/src/types.ts]: [Updated](https://github.com/ChainSafe/web3.js/pull/4875/commits/6c6d829a53e7e46e85a1784b894b756ec6e07cf4), but `SendSignedTransactionEvents` emits `HexStringBytes` for both `sending` and `sent` as opposed to a full `Transaction` object in `SendTransactionEvents

File: packages/web3-eth/src/utils/watch_transaction_for_confirmations.ts (hunk 4 above)
Comment (Spacesai1or) [reply packages/web3-eth/src/utils/watch_transaction_for_confirmations.ts:29]: [Updated](https://github.com/ChainSafe/web3.js/pull/4875/commits/175c6d66debbd177040de8d462738b1784e8f982)

File: packages/web3-eth/src/errors.ts [hunk 5]
```diff
@@ -23,7 +23,7 @@ import {
 	ERR_TX_RECEIPT_MISSING_BLOCK_NUMBER,
 	ReceiptInfo,
 } from 'web3-common';
-import { HexString, HexString32Bytes, Numbers, Web3Error } from 'web3-utils';
+import { HexString, HexString32Bytes, HexStringBytes, Numbers, Web3Error } from 'web3-utils';
```
Comment (Spacesai1or) [inline packages/web3-eth/src/errors.ts]:
```suggestion
import { HexString, HexString32Bytes, Numbers, Web3Error } from 'web3-utils';
```

**Review by jdevcs — COMMENTED** [review]
Good work @spacesailor24 , LGTM, last thing to fix : failing unit tests: `test/unit/rpc_method_wrappers/fixtures/sendTransaction.ts:20:7 - error TS2741: Property 'effectiveGasPrice' is missing in type '{ transactionHash: string; transactionIndex: string; blockHash: string; blockNumber: string; from: string; to: string; cumulativeGasUsed: string; gasUsed: string; contractAddress: string; logs: never[]; logsBloom: string; root: string; status: "0x1"; }' but required in type 'ReceiptInfo'.`

**Comment by Spacesai1or** [comment]
This PR is now includes #4859, and will need to be merged after it is

File: packages/web3-eth/src/rpc_method_wrappers.ts (hunk 1 above)
Comment (nazarhussain) [reply packages/web3-eth/src/rpc_method_wrappers.ts]: I would suggest to keep it `async`, for future extensibility.

**Review by nazarhussain — APPROVED** [review]
LGTM, would be nice to review specific changes once the base PR is merged.

**Review by jdevcs — APPROVED** [review]

**Pull request merged**